### DISCUSSION TITLE
Feature Ideas for C#, GO, Rust

### DISCUSSION THREAD

**Discussion by badamczewski** [badamczewski/PowerUp#6] [Ideas]
**What new and exciting feature ideas would you like to see for existing Languages (C#, GO, Rust)?**

Here's the current feature set per language:

C#

- Decompiles to C#, IL, ASM
- Has Jump Guides
- Has ASM Documentation
- Has Source Code Maps (ASM Code maps to C# source code, IL Code maps to C# source code)
- Has Layouts - any type will be allocated, and field layout will be collected for it, along with extra hidden metadata.
- Has special attributes to run code and print results
- Has Benchmarking (via a special Benchmark Attribute)
- Has Tiered Compilation (can compile code in T0 and T1 at the same time)
- Has PGO Support (via special attribute) that can optimize code based on usage

GO

- Decompiles to GO ASM
- Has Jump Guides
- Has ASM Documentation
- Has Source Code Maps (ASM Code maps to GO source code)

Rust

- Decompiles to ASM
- Has Jump Guides
- Has ASM Documentation
- Has Source Code Maps (ASM Code maps to Rust source code)

It should go without saying but any idea (even a great one) is not going to be automatically a candidate for implementation. I have limited free time and resources so I get to decide what get's implemented, and when :)

**Comment by pictos** [comment]
C#:

- Create an alias for generated member:
When the C# compiler generates classes, fields, etc. it doesn't use a human-friendly name(e.g. `<>DisplayClass0_0`) in some contexts this is very hard to read the generated code and understand it. The solution that I can imagine is to create an Alias for that, can be something constant `DisplayClass0` or dynamic like:

```csharp

// user code

[GenerateNameAlias("FuncInt")]
Func<int> f => () => x+1;
f();

//generated code

[CompileGenerated]
private sealed class FuncInt
{
     // rest of the code
}
```

**Reply by badamczewski** [reply]
This would require from us to grab the decompiled AST and visit it and emit different string values for specific types. It's definitely possible, but it will require a bit of work.

**Reply by badamczewski** [reply]
Here's what I got so far:
![obraz](https://user-images.githubusercontent.com/752380/150032190-f6e74524-040d-4e5b-b773-15209409b21a.png)

Attributes are not supported on this level, so a code rewriter would need to be used before the compilation even takes place and it's an awful lot of work for something so seemingly trivial, so I'm considering a name simplification using a global setting instead.

**Reply by pictos** [reply]
@badamczewski this looks amazing! 😍
thanks for that
